flight/mcu/lib/teensyShared/TimerFSM.cpp

```diff
@@ -58,120 +54,125 @@ rocketFSM::rocketFSM(pointers *ptr) {
-                *rocket_state_ptr_ = STATE_BOOST;
+            // Keeping rocket in FSM_State::STATE_BOOST if time below a certain
+            // threshold
+            if (TIME_I2MS(burn_timer_) < burn_time_thresh_ms) {
+                rocket_state_ = FSM_State::STATE_BOOST;
             }
-            // Forcing rocket to go to STATE_COAST if threshold crossed
+            // Forcing rocket to go to FSM_State::STATE_COAST if threshold
+            // crossed
             else {
-                *rocket_state_ptr_ = STATE_COAST;
+                rocket_state_ = FSM_State::STATE_COAST_PREGNC;
                 // Setting burnout time because we don't otherwise
-                *burnout_time_ptr_ = chVTGetSystemTime();
+                burnout_time_ = chVTGetSystemTime();
             }
 
             break;
 
-        case STATE_BURNOUT_DETECT:
+        case FSM_State::STATE_BURNOUT_DETECT:
             // If the 0 acceleration was too brief, go back to BOOST
             // Serial.println("BURNOUT_DETECT");
             if (*linear_acceleration_ptr_ > coast_thresh) {
-                *rocket_state_ptr_ = STATE_BOOST;
+                rocket_state_ = FSM_State::STATE_BOOST;
                 break;
             }
 
             // Measure the length of the coast time (for hysteresis)
-            *coast_timer_ptr_ = chVTGetSystemTime() - *burnout_time_ptr_;
+            coast_timer_ = chVTGetSystemTime() - burnout_time_;
 
             // If the low acceleration lasts long enough, coast is detected
-            if (TIME_I2MS(*coast_timer_ptr_) > coast_time_thresh) {
-                *rocket_state_ptr_ = STATE_COAST;
+            if (TIME_I2MS(coast_timer_) > coast_time_thresh) {
+                rocket_state_ = FSM_State::STATE_COAST_PREGNC;
             }
 
             break;
 
-        case STATE_COAST:
+        case FSM_State::STATE_COAST_PREGNC:
+            // Serial.println("COAST_PREGNC");
+            // If high acceleration is observed in z direction...
+            coast_timer_ = chVTGetSystemTime() - burnout_time_;
+            if (TIME_I2MS(coast_timer_) > coast_ac_delay_thresh) {
+                rocket_state_ = FSM_State::STATE_COAST_GNC;
+            }
+
+            break;
+
+        case FSM_State::STATE_COAST_GNC:
             // Serial.println("COAST");
-            *coast_timer_ptr_ = chVTGetSystemTime() - *burnout_time_ptr_;
+            coast_timer_ = chVTGetSystemTime() - burnout_time_;
 
-            if (TIME_I2MS(*coast_timer_ptr_) > coast_to_apogee_time_thresh) {
-                *rocket_state_ptr_ = STATE_APOGEE;
+            if (TIME_I2MS(coast_timer_) > coast_to_apogee_time_thresh) {
+                rocket_state_ = FSM_State::STATE_APOGEE;
             } else {
                 // Serial.println("Still in coast");
             }
 
             break;
-        case STATE_APOGEE:
+        case FSM_State::STATE_APOGEE:
         // Serial.println("APOGEE");
```

Comment:
We should delete all the serial prints and just use the THREAD_DEBUG ifdef at the bottom of the tick function to print out what the current state is.